### DISCUSSION TITLE
Debugging devices to aid development

### DISCUSSION THREAD

**Discussion by bytespider** [krahabb/meross_lan#126] [Show and tell]
I've written a tool to help see the messages between Meross devices and the Cloud MQTT server.
[mqtt-debug](https://github.com/bytespider/mqtt-debug)

My hope is by looking at this communication we can help meross_lan do the right thing rather than guessing based on know patterns and observations.

Heres how I inject the tool:
1. Start the pairing process in the official app.
2. When you get to adding the WIFI details, STOP.
3. Connect to the device AP and use a tool like my [meross](https://github.com/bytespider/meross) utilities to query what the currently set broker is and your cloud username and key.
4. HTTP POST to the device the config (see below)
5. Disconnect from AP
6. Start mqtt-debug `mqtt-debug --cert path/to/server-cert --key path/to/server-key --up-cert path/to/client-cert --up-key path/to/client-key` --x-amzn-mqtt-ca 8883 mqtt-eu.meross.com 443`
5. Complete WIFI setup in app.

HTTP POST message:
```
{
	"header": {
		"from": "/app/meross-a13f74f3-a8a7-43c3-adff-adc57edca8c8/",
		"messageId": "5b7e3cd70d5c44c73d9c8e615f4f8686",
		"timestamp": 1637862608,
		"sign": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", // calculate the proper has and populate
		"method": "SET",
		"namespace": "Appliance.Config.Key"
	},
	"payload": {
		"key": {
			"userId": "0000000", // replace with your user id from step 3
			"key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", // replace with your key from step 3
			"gateway": {
				"host": "192.168.1.2", // replace with the IP of the machine you run mqtt-debug from
				"port": "8883",
				"secondHost": "192.168.1.2",  // replace with the IP of the machine you run mqtt-debug from
				"secondPort": "8883"
			}
		}
	}
}

**Comment by krahabb** [comment]
Hello @bytespider ,
this is awesome! I've started thinking of integrating the 'rogue' mqtt into meross_lan to streamline the process but..well it's node.js and I'm not in the mood to port it (not sure it would be wise anyway)

At any rate this would be really helpful. I'll see if I can configure a 'stable' environment in order to plug/unplug it into my own infrastructure

Thank you!

**Reply by bytespider** [reply]
Porting it shouldn't be too hard if you wanted to do that.
Loosely its just a broker one side and a client the other and you replay any message you get one side on the other.

I use node.js just because it's the fastest thing for me to get something built in.

If you use docker, perhaps I could set up a docker container to run this.

**Reply by krahabb** [reply]
Thank you a lot, I was thinking about going the docker way too but I'm totally noob at setting it up: I usually just develop with vscode and it just manages the docker images by itself so I have very very minimal knowledge of it.
Anyway, I've used node with npm too so that shouldn't be  an issue at all configuring and using your tools. I just need to set my mind on this task one of these days but it will not really happen soon
I'd say to not waste your time at this point if it was just for me but thank you very much for the offer ;)

**Reply by bytespider** [reply]
Docker added, and instruction on how to use. For Meross devices, the built in certificates are enough.

```shell
docker run --rm -it -p 8883:8883 mqtt-debug --ca certs/ca.pem --cert certs/cert.pem --key certs/key.pem --up-cert certs/cert.pem --up-key certs/key.pem --x-amzn-mqtt-ca 8883 mqtt-eu.meross.com  443
```
